### PR TITLE
feat: expose descope sdk

### PR DESCRIPTION
**Pull request by omercnet**
## Related Issues

Fixes https://github.com/descope/django-descope/issues/152

## Description

* Add an optional setting to allow using [Descope SDK Management API](https://docs.descope.com/manage/)
* Expose a reusable `descope` client on top level of module
* Fix role validation to use Descope SDK

### CODE REVIEW

**Review by guyp-descope — DISMISSED** [review]

File: django_descope/models.py [hunk 1]
```diff
@@ -3,9 +3,8 @@
 from descope import DescopeClient
 from django.contrib.auth import models as auth_models
 from django.core.cache import cache
-from django.utils.functional import cached_property
 
-from .settings import IS_STAFF_ROLE, IS_SUPERUSER_ROLE, PROJECT_ID
+from .settings import IS_STAFF_ROLE, IS_SUPERUSER_ROLE, MANAGEMENT_KEY, PROJECT_ID
```
Comment (benjamin-kirkbride) [inline django_descope/models.py:7]: should this be `DESCOPE_MANAGEMENT_KEY` and `DESCOPE_PROJECT_ID`?

Comment (benjamin-kirkbride) [reply django_descope/models.py:7]: it seems to work regardless of this actually, but I have no idea why :thinking:

File: django_descope/models.py [hunk 2]
```diff
@@ -14,35 +13,33 @@ class DescopeUser(auth_models.User):
...
     # inactive user
     is_active = True
-    _descope = DescopeClient(PROJECT_ID)
+    descope = DescopeClient(project_id=PROJECT_ID, management_key=MANAGEMENT_KEY)
 
     def sync(self, session, refresh):
-        self.token = session
-        self.session = session.get("jwt")
+        self.session = session
+        self.token = self.session["jwt"]
```
[the remark this answers is not in the document]
Comment (benjamin-kirkbride) [reply django_descope/models.py:23]: This does appear to be nessesary.

File: django_descope/models.py (hunk 1 above)
Comment (omercnet) [reply django_descope/models.py:7]: This actually comes from the app settings and not the project settings

See django_descope/settings.py

File: django_descope/models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (omercnet) [reply django_descope/models.py:23]: This was intentional to point out it's required and should always be there

Did you get a key error thrown at any point for this?

[the remark this answers is not in the document]
Comment (benjamin-kirkbride) [reply django_descope/models.py]: I did get a key error, but I unfortunately didn't save the details. I also got an odd error at one point about the django orm not being able to find a `django-session`. I need to play around more and try to reproduce. I'll switch back to the upstream tonight I hope.

[the remark this answers is not in the document]
Comment (omercnet) [reply django_descope/models.py]: fixed

[the remark this answers is not in the document]
Comment (benjamin-kirkbride) [reply django_descope/models.py]: What was done? I may have just messed something up in my testing, so don't take my word for it that this needed done :sweat_smile:

[the remark this answers is not in the document]
Comment (omercnet) [reply django_descope/models.py]: Check out the latest commit this is outdated

**Review by benjamin-kirkbride — APPROVED** [review]
Looks great! Thanks so much for getting to this so quick :)

File: django_descope/authentication.py [hunk 3]
```diff
@@ -41,14 +39,16 @@ def authenticate(self, request: HttpRequest):
...
+            # Contains sensitive information, so only log in DEBUG mode
+            logger.debug(validated_session)
+        return self.get_user(request, validated_session, refresh_token)
 
-    def get_user(self, request: HttpRequest, validated_token=None, refresh_token=None):
-        if validated_token:
-            username = validated_token.get("userId") or validated_token.get("sub")
+    def get_user(self, request: HttpRequest, validated_session, refresh_token):
+        if validated_session:
+            username = validated_session.get("userId") or validated_session.get("sub")
```
[the remark this answers is not in the document]
Comment (omercnet) [reply django_descope/authentication.py]: fixed

File: django_descope/models.py [hunk 4]
```diff
@@ -14,35 +13,32 @@ class DescopeUser(auth_models.User):
...
     is_active = True
-    _descope = DescopeClient(PROJECT_ID)
 
     def sync(self, session, refresh):
-        self.token = session
-        self.session = session.get("jwt")
-        self.refresh = refresh
-        self.user = session.get("user")
-        self.firstSeen = session.get("firstSeen")
+        self.session_token = session["sessionToken"]  # this should always exist
```
[the remark this answers is not in the document]
Comment (omercnet) [reply django_descope/models.py]: fixed

**Pull request merged**